### PR TITLE
daemon: let client decide whether to allow interactive auth via polkit

### PR DESCRIPTION
**Pull request by jhenstridge**
This is a minor improvement to my `polkit-support` branch that recently got merged: rather than always allowing interactive polkit authorisation, only enable it when the client requests it via a `X-Allow-Interaction: 1` request header.

This branch also updates the `snap` command line tool to set this flag if stdin is a terminal.  This will prevent non-interactive scripts that invoke `snap` from potentially pausing to prompt the user.

This doesn't make much of a difference at the moment since polkit authorisation is only enabled for `snap login`, which also requires interaction.  But if we expand polkit support to more commands, it could become more pressing.

### CODE REVIEW

File: daemon/daemon_test.go [hunk 1]
```diff
@@ -49,11 +50,13 @@ func Test(t *testing.T) { check.TestingT(t) }
 type daemonSuite struct {
 	authorized bool
 	err        error
+	lastFlags  polkit.CheckFlags
```
Comment (mvo5) [inline daemon/daemon_test.go]: (nitpick) Maybe `lastPolkitFlags` ?

**Comment by jhenstridge** [comment]
Just a bit more background for this change: the idea of this HTTP request header is to offer behaviour similar to the D-Bus `ALLOW_INTERACTIVE_AUTHORIZATION` header described here:

https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-messages

The idea is that having an API call involve an arbitrary delay caused by user interaction is unusual behaviour so should only occur if the client says it is okay with it.

File: client/client.go [hunk 2]
```diff
@@ -175,6 +187,10 @@ func (client *Client) raw(method, urlpath string, query url.Values, headers map[
 		}
 	}
 
+	if client.interactive {
+		req.Header.Set(AllowInteractionHeader, "1")
```
Comment (zyga) [inline client/client.go]: I think we traditionally use `true` instead of 1.

File: daemon/daemon.go [hunk 3]
```diff
@@ -101,7 +102,11 @@ func (c *Command) canAccess(r *http.Request, user *auth.UserState) bool {
 		}
 
 		if c.PolkitOK != "" {
-			if authorized, err := polkitCheckAuthorizationForPid(pid, c.PolkitOK, nil, polkit.CheckAllowInteraction); err == nil {
+			var flags polkit.CheckFlags
+			if r.Header.Get(client.AllowInteractionHeader) == "1" {
```
Comment (zyga) [inline daemon/daemon.go]: Same comment about the "true" thing. I think we also may have a helper for that but perhaps I'm confused (helper about boolean things like this).

Comment (chipaca) [reply daemon/daemon.go]: my suggestion here would be to have the client send `"true"`, but use `strconv.ParseBool` in the ademon to accept anything true-ish.

File: daemon/daemon.go [hunk 4]
```diff
@@ -101,7 +103,16 @@ func (c *Command) canAccess(r *http.Request, user *auth.UserState) bool {
 		}
 
 		if c.PolkitOK != "" {
-			if authorized, err := polkitCheckAuthorizationForPid(pid, c.PolkitOK, nil, polkit.CheckAllowInteraction); err == nil {
+			allow, err := strconv.ParseBool(r.Header.Get(client.AllowInteractionHeader))
```
Comment (niemeyer) [inline daemon/daemon.go]: This seems to be failing and reporting an error on an empty header. It should instead silently use the default behavior.

File: daemon/daemon.go [hunk 5]
```diff
@@ -105,7 +105,8 @@ func (c *Command) canAccess(r *http.Request, user *auth.UserState) bool {
 		if c.PolkitOK != "" {
 			allow, err := strconv.ParseBool(r.Header.Get(client.AllowInteractionHeader))
 			if err != nil {
-				logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
+				// default behaviour if the header
+				// cannot be parsed
```
Comment (niemeyer) [inline daemon/daemon.go]: @mvo5 That is ignoring _all_ errors, including improper headers. The logging as done in the original PR seems fine for cases which are not related to a missing header. I believe the Get will return "" in those.

**Comment by robert-ancell** [comment]
API landed in snapd-glib 1.19 and the snapd REST API documentation is updated.

**Pull request merged**